File: 1_Two_Sum/cpp/main.cpp
#include <iostream>
#include "TwoSum.h"

int main()
{
    TwoSum solution;

    // Test case 1
    std::vector<int> nums1{2, 7, 11, 15};
    int target1 = 9;
    std::vector<int> result1 = solution.twoSum(nums1, target1);
    std::cout << "Test case 1: ";
    if (result1.size() == 2 && result1[0] == 0 && result1[1] == 1)
    {
        std::cout << "Passed\n";
    }
    else
    {
        std::cout << "Failed\n";
    }

    // Test case 2
    std::vector<int> nums2{3, 2, 4};
    int target2 = 6;
    std::vector<int> result2 = solution.twoSum(nums2, target2);
    std::cout << "Test case 2: ";
    if (result2.size() == 2 && result2[0] == 1 && result2[1] == 2)
    {
        std::cout << "Passed\n";
    }
    else
    {
        std::cout << "Failed\n";
    }

    // Test case 3
    std::vector<int> nums3{3, 3};
    int target3 = 6;
    std::vector<int> result3 = solution.twoSum(nums3, target3);
    std::cout << "Test case 3: ";
    if (result3.size() == 2 && result3[0] == 0 && result3[1] == 1)
    {
        std::cout << "Passed\n";
    }
    else
    {
        std::cout << "Failed\n";
    }

    return 0;
}

File: 1_Two_Sum/rust/src/main.rs
use std::collections::HashMap;

struct Solution {}

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        // Create a new HashMap
        let mut map = HashMap::new();
        // Iterate over the vector of numbers
        for (i, num) in nums.iter().enumerate() {
            // Find the complement of the current value
            let complement = target - num;
            // If the complement exists in the HashMap, return the current index and the index of the complement
            if map.contains_key(&complement) {
                // Return a vector of the current index and the index of the complement
                return vec![map[&complement], i as i32];
            }
            // Add the current value and its index to the HashMap
            map.insert(num, i as i32);
        }
        // If no two numbers add up to the target, return an empty vector
        vec![]
    }
}

fn main() {
    let nums1 = vec![2, 7, 11, 15];
    let target1 = 9;
    assert_eq!(Solution::two_sum(nums1, target1), vec![0, 1]);

    let nums2 = vec![3, 2, 4];
    let target2 = 6;
    assert_eq!(Solution::two_sum(nums2, target2), vec![1, 2]);

    let nums3 = vec![3, 3];
    let target3 = 6;
    assert_eq!(Solution::two_sum(nums3, target3), vec![0, 1]);
}
